src/thread.cr

```diff
@@ -63,8 +68,10 @@ class Thread
 
       current_thread_id = LibC.pthread_self
 
-      current_thread = @@threads.find do |thread|
-        LibC.pthread_equal(thread.id, current_thread_id) != 0
+      current_thread = @@threads.each_key do |thread|
+        if LibC.pthread_equal(thread.id, current_thread_id) != 0
+          break thread
+        end
```

Comment:
Overloading `#hash(hasher)` is simpler and avoids additional types. If it achieves the same, let's go for it. I doubt it could be an attack vector for `Thread` or `Signal`, and they're only supposed to be used internally.

Note: did you really mean `Enum#hash`?